### ISSUE TITLE
sailfish/marlin: Can't use sudo apt-get update to update software

### ISSUE THREAD

**Issue by utzcoz** [maruos/maruos#128]
Use `lxc-start --name default` in `shell` to start `debian`. After authentication, use `sudo apt-get update` to update software resource, the command show below error messages:

```
Reading package lists... Error!
E: Could not open lock file /var/lib/apt/lists/lock - open (126: Required key not available)
E: Unable to lock directory /var/lib/apt/lists/
E: Could not create temporary file for /var/cache/apt/srcpkgcache.bin - mkstemp (126: Required key not available)
E: The package lists or status file could not be parsed or opened.
```

After that, I try to execute `sudo apt-get update` again, it show below error message:

```
sudo: unable to open /var/lib/sudo/ts/maru: Operation not permitted
[sudo] password for maru: 
Reading package lists... Error!
E: Could not open lock file /var/lib/apt/lists/lock - open (126: Required key not available)
E: Unable to lock directory /var/lib/apt/lists/
E: Could not create temporary file for /var/cache/apt/srcpkgcache.bin - mkstemp (126: Required key not available)
E: The package lists or status file could not be parsed or opened.
```

And I found below error messages from `demsg`:

```
[   69.848847] EXT4-fs (sda32): errors=remount-ro for active namespaces on umount 0
[   69.849267] EXT4-fs (sdd3): errors=remount-ro for active namespaces on umount 0
[   69.852489] EXT4-fs (sda35): errors=remount-ro for active namespaces on umount 0
[   69.854777] EXT4-fs (dm-0): errors=remount-ro for active namespaces on umount 2
```

The `sda35` is the `data` partition of Android.

After `lxc-start --name default`, the mount `errors` parameters of `/data` of Android has been changed to `remount-ro` from `panic` that I set from `fstab`.

**Comment by utzcoz** [comment]
I open the `CONFIG_EXT4_DEBUG` in `defconfig`, and use `echo 1 > /sys/module/ext4/parameters/mballoc_debug` to open the `ext4` runtime debug, and I found below log before `EXT4-fs (sda35)`:

```
[ 1107.244763] (maruos/kernel/google/marlin/fs/ext4/mballoc.c, 3942): ext4_discard_preallocations: discard preallocation for inode 917517
[ 1107.244969] (maruos/kernel/google/marlin/fs/ext4/mballoc.c, 3942): ext4_discard_preallocations: discard preallocation for inode 917517
[ 1107.250815] (maruos/kernel/google/marlin/fs/ext4/mballoc.c, 4198): ext4_mb_initialize_context: init ac: 1 blocks @ 0, goal 4194304, flags 0, 2^0, left: 0/0, right 0/0 to non-writable
[ 1107.250862] (maruos/kernel/google/marlin/fs/ext4/mballoc.c, 1115): ext4_mb_load_buddy: load group 128
```

Those logs are printed when I use `lxc-start --name default` to start `debian`.

**Comment by utzcoz** [comment]
For `sudo: unable to open /va/lib/sudo/ts/maru: Operaiton not permitted`, I have found a page to describe it, [Debian Bug report logs - #712839 sudo: /var/lib/sudo permissions](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=712839). The page thinks its a `debian` `postinit` bug with wrong permission.

**Comment by pdsouza** [comment]
@utzcoz Have you tried deleting `/var/lib/sudo` and re-creating with correct permissions?

**Comment by utzcoz** [comment]
@pdsouza After deleting `/var/lib/sudo`,  and using `sudo apt-get update`, the `/var/lib/sudo` is created with permission `711`. And the directory `/var/lib/sudo` is empty.

**Comment by utzcoz** [comment]
I check the directory `/var/lib/sudo` in my `Ubuntu` laptop, and it has permission `755`. If I try to use `sudo mkdir ts` to create the `ts` directory, the `ts` directory gets the permission `755`.

**Comment by TMartinPPC** [comment]
I could be wrong but I don't think that ts is a directory. It's not on my Debian install on my Chromebook.

**Comment by utzcoz** [comment]
@pdsouza @TMartinPPC  I use `blueprints` to build a `Debian buster` rootfs, and flash it to system. The error `sudo: unable to open /va/lib/sudo/ts/maru: Operaiton not permitted` disappeared. And `sudo` won't try to create `ts` directory in `/var/lib/sudo`, likely `Debian` in @TMartinPPC Chromebook. And I have installed `Debian buster` in `virtualbox`, and its `sudo` won't try to create `ts` directory in `/var/lib/sudo` too.

**Comment by utzcoz** [comment]
Not only `apt get update`, when I use `mktemp` to create temp file, the command prints the similar error:
```
mktemp: failed to create file via template ‘/tmp/tmp.XXXXXXXXXX’: Required key not available
```
Then I rebuild rootfs with `strace` binary, and use `strace mktemp` to show the detail of command executing, and get below result:
```
execve("/bin/mktemp", ["mktemp"], 0x7feb348e70 /* 22 vars */) = 0
brk(NULL)                               = 0x59040b8000
faccessat(AT_FDCWD, "/etc/ld.so.preload", R_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=27181, ...}) = 0
mmap(NULL, 27181, PROT_READ, MAP_PRIVATE, 3, 0) = 0x77c024e000
close(3)                                = 0
openat(AT_FDCWD, "/lib/aarch64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0\267\0\1\0\0\0p\16\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1439544, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x77c024c000
mmap(NULL, 1511520, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x77c00b6000
mprotect(0x77c020f000, 61440, PROT_NONE) = 0
mmap(0x77c021e000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x158000) = 0x77c021e000
mmap(0x77c0224000, 12384, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x77c0224000
close(3)                                = 0
mprotect(0x77c021e000, 16384, PROT_READ) = 0
mprotect(0x58cd8c8000, 4096, PROT_READ) = 0
mprotect(0x77c0257000, 4096, PROT_READ) = 0
munmap(0x77c024e000, 27181)             = 0
brk(NULL)                               = 0x59040b8000
brk(0x59040d9000)                       = 0x59040d9000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=3031632, ...}) = 0
mmap(NULL, 3031632, PROT_READ, MAP_PRIVATE, 3, 0) = 0x77bfdd1000
close(3)                                = 0
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 3
read(3, "\241LW\230\272\300!~\6\221", 10) = 10
close(3)                                = 0
getpid()                                = 796
getppid()                               = 793
getuid()                                = 901000
getgid()                                = 901000
openat(AT_FDCWD, "/tmp/tmp.XXiQqO4rAU", O_RDWR|O_CREAT|O_EXCL, 0600) = -1 ENOKEY (Required key not available)
openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2995, ...}) = 0
read(3, "# Locale name alias data base.\n#"..., 4096) = 2995
read(3, "", 4096)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale/en_US.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en_US.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en_US/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, "mktemp: ", 8mktemp: )                 = 8
write(2, "failed to create file via templa"..., 60failed to create file via template ‘/tmp/tmp.XXXXXXXXXX’) = 60
openat(AT_FDCWD, "/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, ": Required key not available", 28: Required key not available) = 28
write(2, "\n", 1
)                       = 1
close(1)                                = 0
close(2)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```
An I found this log: 
```
openat(AT_FDCWD, "/tmp/tmp.XXiQqO4rAU", O_RDWR|O_CREAT|O_EXCL, 0600) = -1 ENOKEY (Required key not available)
```
The call of `openat` throws a `ENOKEY`, but I dont' find the case to throw `ENOKEY` in [Debian openat(2) manpage](https://manpages.debian.org/buster/manpages-dev/openat.2.en.html).

**Comment by utzcoz** [comment]
After searching, many article says the `Required key not available` is because of kernel module signature. But I get below result with command `zcat /proc/config.gz | grep _SIG`:
```
CONFIG_SIGNALFD=y
CONFIG_OLD_SIGSUSPEND3=y
CONFIG_COMPAT_OLD_SIGACTION=y
# CONFIG_MODULE_SIG is not set
# CONFIG_SIGNED_PE_FILE_VERIFICATION is not set
```
The `CONFIG_MODULE_SIG` is disabled on device.

**Comment by utzcoz** [comment]
After enabling log in `kernel/security/keys`, I got some related log from `dmesg` after `sudo apt-get update`:
```
[  331.246612] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.246718] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.246727] descend to 301848468
[  331.246736] begin_node
[  331.246741] not_this_keyring 0
[  331.246746] <== search_nested_keyrings() = false
[  331.246753] <== request_key_and_link() = 0000000000000000
[  331.246790] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.246797] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.246802] descend to 301848468
[  331.246807] begin_node
[  331.246811] not_this_keyring 0
[  331.246816] <== search_nested_keyrings() = false
[  331.246821] <== request_key_and_link() = 0000000000000000
[  331.252020] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.252049] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.252056] descend to 301848468
[  331.252065] begin_node
[  331.252070] not_this_keyring 0
[  331.252075] <== search_nested_keyrings() = false
[  331.252082] <== request_key_and_link() = 0000000000000000
[  331.252094] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.252101] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.252105] descend to 301848468
[  331.252110] begin_node
[  331.252114] not_this_keyring 0
[  331.252119] <== search_nested_keyrings() = false
[  331.252124] <== request_key_and_link() = 0000000000000000
[  331.252149] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.252156] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.252161] descend to 301848468
[  331.252166] begin_node
[  331.252170] not_this_keyring 0
[  331.252175] <== search_nested_keyrings() = false
[  331.252180] <== request_key_and_link() = 0000000000000000
[  331.252188] ==> request_key_and_link(logon,ext4:d9530d41e63740e3,0000000000000000,0,0000000000000000,0000000000000000,0)
[  331.252193] ==> search_nested_keyrings({301848468},{logon,ext4:d9530d41e63740e3})
[  331.252198] descend to 301848468
[  331.252202] begin_node
[  331.252207] not_this_keyring 0
[  331.252211] <== search_nested_keyrings() = false
[  331.252216] <== request_key_and_link() = 0000000000000000
```

**Comment by utzcoz** [comment]
If we want to get above log, we should enable `CONFIG_KEYS_DEBUG_PROC_KEYS` in defconfig.

**Comment by utzcoz** [comment]
If use `sudo cat /proc/keys` in Linux, it will get the same result with `cat /proc/keys` in Android for root user.

**Comment by pdsouza** [comment]
@utzcoz Thanks for continuing to debug this and sharing your logs! It looks like all of these issues stem from `mktemp`/`mkstemp` throwing `ENOKEY`. This is a very strange...I wonder why `/tmp` is having these errors. 

Some things to try:
* What happens if you `touch` a file in `/tmp` without using `mktemp`?
* Is there anything strange about the `tmpfs` mount in `fstab`? Are there any kernel options enabled on `sailfish` that affect `tmpfs`?

**Comment by utzcoz** [comment]
@pdsouza `touch temp` works fine in `/tmp`. `ENOKEY` represents the process doesn't find required key. What I did before is to check whether `Android` and `Linux` can get the same key list. From the result, they can get the same list.

**Comment by utzcoz** [comment]
I compare the input parameter of `search_nested_keyrings`, and found that when I execute `mktemp` in `Android` shell, the log showed the `search_nested_keyrings` with the input serial of key, but when for `Linux` shell, the log showed the `search_nested_keyrings` with the wrong input serial of key. And the key serail is got from `struct cred`, created by `current_cred()` method.

**Comment by utzcoz** [comment]
The `ENOKEY` is for the `session_keyring`. And from the kernel document, `session_kerying` is from the parent process, when it forked from parent process. Compare the `search_nested_keyrings` input serial and `cat /proc/keys` result from `Linux` shell, I found that `search_nested_keyrings` input serial is the a session keyring viewed by `cat /proc/keys`. So it looks like `Linux` can't get correct key list contains the key requested by program.

**Comment by utzcoz** [comment]
@pdsouza The problem is that all above commands are executed in `Linux` started with `lxc-start --name default` in `adb shell`. I use `busybox pstree` to see the process relationship, and all `Linux` processes are inherited from `adbd`. So all `Linux` process inherited session keyring from `adbd`.  It doesn't work in `Linux` occasion. If we start `Linux` from `perspectived`, all things are correct. I can use `apt-get update` to update apt list. So `maruos` doesn't work correctly when we use `lxc-start --name default` to start `Linux` in `adb shell` in `pixel`. So I close the issue.

**Comment by pdsouza** [comment]
@utzcoz Wow, very interesting. It is good you have documented your debugging in this issue for other users.

**Comment by utzcoz** [comment]
I should correct the parent process of `Linux` process, they are forked from `init` not `perspectived`.

```
marlin:/ # ps -A | grep systemd
root          3655  3609   25096   4872 SyS_epoll_wait 702ac78bb0 S systemd-journald
root          3678  3609   17628   2852 SyS_epoll_wait 75f2872bb0 S systemd-udevd
root          3724  3609   14768   3876 SyS_epoll_wait 75338d8bb0 S systemd-logind
marlin:/ # ps -A | grep 3609
root          3609  3594   17356   6400 SyS_epoll_wait 71f202cbb0 S init
root          3655  3609   25096   4872 SyS_epoll_wait 702ac78bb0 S systemd-journald
root          3678  3609   17628   2852 SyS_epoll_wait 75f2872bb0 S systemd-udevd
root          3724  3609   14768   3876 SyS_epoll_wait 75338d8bb0 S systemd-logind
root          3725  3609  391444   6680 poll_schedule_timeout 7eb99212f4 S udisksd
104           3728  3609    7088   2288 SyS_epoll_wait 7efae18bb0 S dbus-daemon
root          3746  3609    5068    744 poll_schedule_timeout 7cc6db7d3c S agetty
root          3750  3609  232320   3756 poll_schedule_timeout 798d58d2f4 S polkitd
root          9469  3609  305944   4112 poll_schedule_timeout 7cdc4f32f4 S lightdm
```

**Comment by utzcoz** [comment]
@pdsouza I test this problem again, and I can use `lxc-start --name default`  and open simulate second display  to show the `Linux` to display correctly when using `Debian stretch`, but not for `Debian buster`. So the previous conclusion should be changed.
When I use `lxc-start --name default` to start `Linux` in `adbd`, the `adbd` has the `root` priviledge. So `lxc` processors and its forked processor maybe the same as processors forked from `perspectived`.  But `Linux` can't use `apt-get update` to update the software list in `stretch` and `buster`.
And I read the `keys.txt` documentation in `Linux` `kernel` again, I found some useful information. Above log shows the `apt-get update` failed with to request key `ext4:786a7fe8205b62eb`, and its information showed by `cat /proc/keys` is:
```
2e1e0a9c I--Q---     1 perm 3d010000     0     0 logon     ext4:786a7fe8205b62eb: 72
```
This key owner user is `0` and group is `0`. The `logon` represents this key should only be used by `kernel` not allowed for `userspace`. To get more detail, I think I should inspect it again, so I reopen the issue to track this problem.

**Comment by utzcoz** [comment]
And there is a similar issue [ required key not available in docker container #128](https://github.com/google/fscrypt/issues/128.

**Comment by utzcoz** [comment]
Okay, the newest result is that: 
The `buster` uses the `systemd` `v241`, and `stretch` uses the `systemd` `v232`, what we can get from `debian` package website. From the version `v232` to `v241`, the `systemd` add a patch [core: run each system service with a fresh session keyring](https://github.com/systemd/systemd/commit/74dd6b515fa968c5710b396a7664cac335e25ca8), that calls `keyctl(KEYCTL_JOIN_SESSION_KEYRING, 0, 0, 0, 0)` to create new session keyring for each service. So the `systemd-sysusers` will get an anonymous session. But it will try to fetch some kernel keyring, and fail because of the `Required key not found` error. After it failed, the `maru-mflinger-client` maybe get incorrect envrionment, so it started failed. So I try to hack to disable `KEYCTL_JOIN_SESSION_KEYRING` in `marlin` kernel, and it works fine. After hacking, `Linux` shows correctly in `simulate secondary display`. And I can use `sudo apt-get update` and `sudo apt-get install emacs` successfully.
The hack detail is in the [pr #2](https://github.com/maruos/android_kernel_google_marlin/pull/2).
So I close this issue after hack merging.
